src/Dfc.CourseDirectory.Web/Controllers/CopyCourse/CopyCourseRunController.cs

```diff
@@ -262,243 +219,355 @@ private async Task<SelectVenueModel> GetVenuesByUkprn(int ukprn)
         {
-            int? UKPRN;
+            if (!_session.GetInt32("UKPRN").HasValue)
+            {
+                return RedirectToAction("Index", "Home", new { errmsg = "Please select a Provider." });
+            }
+
+            var ukprn = _session.GetInt32("UKPRN").Value;
+
+            var savedModel = _session.GetObject<CopyCourseRunSaveViewModel>(CopyCourseRunSaveViewModelSessionKey);
 
-            if (_session.GetInt32("UKPRN") != null)
+            if (courseId.HasValue && courseRunId.HasValue)
+            {
+                if (savedModel != null)
+                {
+                    if (savedModel.CourseId != courseId || savedModel.CourseRunId != courseRunId)
+                    {
+                        _session.Remove(CopyCourseRunSaveViewModelSessionKey);
+                        savedModel = null;
+                    }
+                    else
+                    {
+                        courseId = savedModel.CourseId;
+                        courseRunId = savedModel.CourseRunId;
+                    }
+                }
+            }
+            else if (savedModel != null)
             {
-                UKPRN = _session.GetInt32("UKPRN").Value;
+                courseId = savedModel.CourseId;
+                courseRunId = savedModel.CourseRunId;
             }
             else
             {
-                return RedirectToAction("Index", "Home", new { errmsg = "Please select a Provider." });
+                return View("Error", new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
             }
 
-            List<SelectListItem> courseRunVenues = new List<SelectListItem>();
+            var course = await _courseService.GetCourseByIdAsync(new GetCourseByIdCriteria(courseId.Value));
 
-            VenueSearchCriteria criteria = new VenueSearchCriteria(UKPRN.ToString(), null);
-            var venues = await _venueService.SearchAsync(criteria);
+            if (!course.HasValue)
+            {
+                return View("Error", new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
+            }
 
-            foreach (var venue in venues.Value.Value)
+            var courseRun = course.Value.CourseRuns.SingleOrDefault(cr => cr.id == courseRunId);
+
+            if (courseRun == null)
             {
-                var item = new SelectListItem { Text = venue.VenueName, Value = venue.ID };
-                courseRunVenues.Add(item);
-            };
+                return View("Error", new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
+            }
 
+            var venues = await _venueService.SearchAsync(new VenueSearchCriteria(ukprn.ToString(), null));
             var regions = _courseService.GetRegions();
 
-            if (courseId.HasValue)
+            foreach (var subRegion in regions?.RegionItems.SelectMany(r => r.SubRegion))
```

Comment:
Could NullRef here if `regions` is `null`

Reply:
`Regions` should never be `null`, so the use of the null conditional operator here is misleading.